<!-- multi_camera_node.launch -->
<!-- Launch a "multi_camera_node" node from "avt_vimba_camera" to get data from multiple cameras, using only one node -->
<!-- Created by pointlaz -->

<launch>
  <!-- multi_camera_node parameters, these parameters define the cameras parameters -->
  <!-- Change these parameters using the quantity and ID of your cameras -->

  <!-- Quantity of cameras -->
  <arg name="camera_qty"                default="7" />
  <!-- Camera 0 -->
  <arg name="guid_0"                    default="DEV_1AB22C00D5C3"/>
  <arg name="name_0"                    default="camera_0"/>
  <arg name="frame_id_0"                default="camera_0"/>
  <arg name="camera_info_url_0"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 1 -->
  <arg name="guid_1"                    default="DEV_1AB22C00D5C7"/>
  <arg name="name_1"                    default="camera_1"/>
  <arg name="frame_id_1"                default="camera_1"/>
  <arg name="camera_info_url_1"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 2 -->
  <arg name="guid_2"                    default="DEV_1AB22C00D5C4"/>
  <arg name="name_2"                    default="camera_2"/>
  <arg name="frame_id_2"                default="camera_2"/>
  <arg name="camera_info_url_2"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 3 -->
  <arg name="guid_3"                    default="DEV_1AB22C00D5C9"/>
  <arg name="name_3"                    default="camera_3"/>
  <arg name="frame_id_3"                default="camera_3"/>
  <arg name="camera_info_url_3"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 4 -->
  <arg name="guid_4"                    default="DEV_1AB22C00D5C5"/>
  <arg name="name_4"                    default="camera_4"/>
  <arg name="frame_id_4"                default="camera_4"/>
  <arg name="camera_info_url_4"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 5 -->
  <arg name="guid_5"                    default="DEV_1AB22C00D5C2"/>
  <arg name="name_5"                    default="camera_5"/>
  <arg name="frame_id_5"                default="camera_5"/>
  <arg name="camera_info_url_5"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>
  <!-- Camera 6 -->
  <arg name="guid_6"                    default="DEV_1AB22C00D5C6"/>
  <arg name="name_6"                    default="camera_6"/>
  <arg name="frame_id_6"                default="camera_6"/>
  <arg name="camera_info_url_6"         default="/home/pointlaz/catkin_ws/src/avt_vimba_camera/calibrations/calibration_example.yaml"/>

  <!-- Driver parameters, these parameters affect the way the ROS driver works -->
  <arg name="use_measurement_time"      default="false"   doc="Use the measurement time from the camera"/>
  <arg name="ptp_offset"                default="0"       doc="Offset to add to the timestamp from the camera (s). Useful for converting from TAI to UTC time. Only applies if use_measurement_time is true"/>
  <arg name="camera_info_url"           default="file://$(find avt_vimba_camera)/calibrations/calibration_example.yaml"/>
  <arg name="print_all_features"        default="false"   doc="Show all camera features on the console during startup"/>
  <arg name="image_proc"                default="false"   doc="Launch image_proc with driver"/>

  <!-- Camera parameters, these parameters are used to configure the AVT camera upon startup -->
  <!-- See the cfg/AvtVimbaCamera.cfg file for documentation -->

  <!-- Acquisition -->
  <arg name="acquisition_mode"          default="Continuous"/>
  <arg name="acquisition_rate"          default="10"/>
  <!-- Trigger -->
  <arg name="trigger_source"            default="Line1"/>
  <arg name="trigger_mode"              default="On"/>
  <arg name="trigger_selector"          default="FrameStart"/>
  <arg name="trigger_activation"        default="RisingEdge"/>
  <arg name="trigger_delay"             default="0.0"/>
  <!-- Exposure -->
  <arg name="exposure"                  default="10000"/>
  <arg name="exposure_auto"             default="Off"/>
  <arg name="exposure_mode"             default="Timed"/>
  <arg name="exposure_auto_alg"         default="FitRange"/>
  <arg name="exposure_auto_tol"         default="5"/>
  <arg name="exposure_auto_max"         default="20000"/>
  <arg name="exposure_auto_min"         default="1000"/>
  <arg name="exposure_auto_outliers"    default="10000"/>
  <arg name="exposure_auto_rate"        default="10000"/>
  <arg name="exposure_auto_target"      default="10000"/>
  <!-- Gain -->
  <arg name="gain"                      default="0"/>
  <arg name="gain_auto"                 default="Off"/>
  <arg name="gain_auto_adjust_tol"      default="5"/>
  <arg name="gain_auto_max"             default="32"/>
  <arg name="gain_auto_min"             default="0"/>
  <arg name="gain_auto_outliers"        default="0"/>
  <arg name="gain_auto_rate"            default="50"/>
  <arg name="gain_auto_target"          default="50"/>
  <!-- White Balance -->
  <arg name="balance_ratio_abs"         default="1.0"/>
  <arg name="balance_ratio_selector"    default="Red"/>
  <arg name="whitebalance_auto"         default="Off"/>
  <arg name="whitebalance_auto_tol"     default="5"/>
  <arg name="whitebalance_auto_rate"    default="100"/>
    <!-- Binning and Decimation -->
  <arg name="binning_x"                 default="1"/>
  <arg name="binning_y"                 default="1"/>
  <arg name="decimation_x"              default="1"/>
  <arg name="decimation_y"              default="1"/>
  <!-- ROI -->
  <arg name="width"                     default="1920"/>
  <arg name="height"                    default="1440"/>
  <arg name="offset_x"                  default="0"/>
  <arg name="offset_y"                  default="0"/>
  <!-- Pixel Format -->
  <arg name="pixel_format"              default="BayerRG8"/>
  <!-- Bandwidth -->
  <arg name="stream_bytes_per_second"   default="50000000"/>
  <!-- PTP Synchronization -->
  <arg name="ptp_mode"                  default="Off"/>
  <!-- USB GPIO -->
  <arg name="line_selector"             default="Line0"/>
  <arg name="line_mode"                 default="Input"/>

  <group if="$(arg image_proc)" ns="$(arg name)">
    <node name="image_proc" pkg="image_proc" type="image_proc"/>
  </group>

  <!-- Run the multi_camera_node -->
  <node name="multi_camera" pkg="avt_vimba_camera" type="multi_camera_node" output="screen">
    <!-- Quantity of cameras -->
    <param name="camera_qty"                value="$(arg camera_qty)" />
    <!-- Camera 0 -->
    <param name="guid_0"                    value="$(arg guid_0)"/>
    <param name="name_0"                    value="$(arg name_0)"/>
    <param name="frame_id_0"                value="$(arg frame_id_0)"/>
    <param name="camera_info_url_0"         value="$(arg camera_info_url_0)"/>
    <!-- Camera 1 -->
    <param name="guid_1"                    value="$(arg guid_1)"/>
    <param name="name_1"                    value="$(arg name_1)"/>
    <param name="frame_id_1"                value="$(arg frame_id_1)"/>
    <param name="camera_info_url_1"         value="$(arg camera_info_url_1)"/>
    <!-- Camera 2 -->
    <param name="guid_2"                    value="$(arg guid_2)"/>
    <param name="name_2"                    value="$(arg name_2)"/>
    <param name="frame_id_2"                value="$(arg frame_id_2)"/>
    <param name="camera_info_url_2"         value="$(arg camera_info_url_2)"/>
    <!-- Camera 3 -->
    <param name="guid_3"                    value="$(arg guid_3)"/>
    <param name="name_3"                    value="$(arg name_3)"/>
    <param name="frame_id_3"                value="$(arg frame_id_3)"/>
    <param name="camera_info_url_3"         value="$(arg camera_info_url_3)"/>
    <!-- Camera 4 -->
    <param name="guid_4"                    value="$(arg guid_4)"/>
    <param name="name_4"                    value="$(arg name_4)"/>
    <param name="frame_id_4"                value="$(arg frame_id_4)"/>
    <param name="camera_info_url_4"         value="$(arg camera_info_url_4)"/>
    <!-- Camera 5 -->
    <param name="guid_5"                    value="$(arg guid_5)"/>
    <param name="name_5"                    value="$(arg name_5)"/>
    <param name="frame_id_5"                value="$(arg frame_id_5)"/>
    <param name="camera_info_url_5"         value="$(arg camera_info_url_5)"/>
    <!-- Camera 6 -->
    <param name="guid_6"                    value="$(arg guid_6)"/>
    <param name="name_6"                    value="$(arg name_6)"/>
    <param name="frame_id_6"                value="$(arg frame_id_6)"/>
    <param name="camera_info_url_6"         value="$(arg camera_info_url_6)"/>
    <!-- Driver parameters -->
    <param name="use_measurement_time"      value="$(arg use_measurement_time)"/>
    <param name="ptp_offset"                value="$(arg ptp_offset)"/>
    <param name="camera_info_url"           value="$(arg camera_info_url)"/>
    <param name="print_all_features"        value="$(arg print_all_features)"/>
    <!-- Acquisition -->
    <param name="acquisition_mode"          value="$(arg acquisition_mode)"/>
    <param name="acquisition_rate"          value="$(arg acquisition_rate)"/>
    <!-- Trigger -->
    <param name="trigger_source"            value="$(arg trigger_source)"/>
    <param name="trigger_mode"              value="$(arg trigger_mode)"/>
    <param name="trigger_selector"          value="$(arg trigger_selector)"/>
    <param name="trigger_activation"        value="$(arg trigger_activation)"/>
    <param name="trigger_delay"             value="$(arg trigger_delay)"/>
    <!-- Exposure -->
    <param name="exposure"                  value="$(arg exposure)"/>
    <param name="exposure_mode"             value="$(arg exposure_mode)"/>
    <param name="exposure_auto"             value="$(arg exposure_auto)"/>
    <param name="exposure_auto_alg"         value="$(arg exposure_auto_alg)"/>
    <param name="exposure_auto_tol"         value="$(arg exposure_auto_tol)"/>
    <param name="exposure_auto_max"         value="$(arg exposure_auto_max)"/>
    <param name="exposure_auto_min"         value="$(arg exposure_auto_min)"/>
    <param name="exposure_auto_outliers"    value="$(arg exposure_auto_outliers)"/>
    <param name="exposure_auto_rate"        value="$(arg exposure_auto_rate)"/>
    <param name="exposure_auto_target"      value="$(arg exposure_auto_target)"/>
    <!-- Gain -->
    <param name="gain"                      value="$(arg gain)"/>
    <param name="gain_auto"                 value="$(arg gain_auto)"/>
    <param name="gain_auto_adjust_tol"      value="$(arg gain_auto_adjust_tol)"/>
    <param name="gain_auto_max"             value="$(arg gain_auto_max)"/>
    <param name="gain_auto_min"             value="$(arg gain_auto_min)"/>
    <param name="gain_auto_outliers"        value="$(arg gain_auto_outliers)"/>
    <param name="gain_auto_rate"            value="$(arg gain_auto_rate)"/>
    <param name="gain_auto_target"          value="$(arg gain_auto_target)"/>
    <!-- White Balance -->
    <param name="balance_ratio_abs"         value="$(arg balance_ratio_abs)"/>
    <param name="balance_ratio_selector"    value="$(arg balance_ratio_selector)"/>
    <param name="whitebalance_auto"         value="$(arg whitebalance_auto)"/>
    <param name="whitebalance_auto_tol"     value="$(arg whitebalance_auto_tol)"/>
    <param name="whitebalance_auto_rate"    value="$(arg whitebalance_auto_rate)"/>
    <!-- Binning and Decimation -->
    <param name="binning_x"                 value="$(arg binning_x)"/>
    <param name="binning_y"                 value="$(arg binning_y)"/>
    <param name="decimation_x"              value="$(arg decimation_x)"/>
    <param name="decimation_y"              value="$(arg decimation_y)"/>
    <!-- ROI -->
    <param name="width"                     value="$(arg width)"/>
    <param name="height"                    value="$(arg height)"/>
    <param name="offset_x"                  value="$(arg offset_x)"/>
    <param name="offset_y"                  value="$(arg offset_y)"/>
    <!-- Pixel Format -->
    <param name="pixel_format"              value="$(arg pixel_format)"/>
    <!-- Bandwidth -->
    <param name="stream_bytes_per_second"   value="$(arg stream_bytes_per_second)"/>
    <!-- PTP Synchronization -->
    <param name="ptp_mode"                  value="$(arg ptp_mode)"/>
    <!-- USB GPIO -->
    <param name="line_selector"             value="$(arg line_selector)"/>
    <param name="line_mode"                 value="$(arg line_mode)"/>

  </node>
</launch>

